packages/account/src/providers/provider.ts

```diff
@@ -771,29 +770,153 @@ export default class Provider {
+  ): Promise<EstimateTxDependenciesReturns[]> {
+    const results: EstimateTxDependenciesReturns[] = transactionRequests.map(() => ({
+      receipts: [],
+      outputVariables: 0,
+      missingContractIds: [],
+      dryrunStatus: undefined,
+    }));
+
+    const allRequests = clone(transactionRequests);
+
+    // Map of original request index to its serialized transaction (for ScriptTransactionRequest only)
+    const serializedTransactionsMap = new Map();
+
+    // Prepare ScriptTransactionRequests and their indices
+    allRequests.forEach((req, index) => {
+      if (req.type === TransactionType.Script) {
+        serializedTransactionsMap.set(index, hexlify(req.toTransactionBytes()));
+      }
+    });
+
+    // Indices of ScriptTransactionRequests
+    let transactionsToProcess = Array.from(serializedTransactionsMap.keys());
+    let attempt = 0;
+
+    while (transactionsToProcess.length > 0 && attempt < MAX_RETRIES) {
+      const encodedTransactions = transactionsToProcess.map((index) =>
+        serializedTransactionsMap.get(index)
+      );
+      const dryRunResults = await this.operations.dryRun({
+        encodedTransactions,
+        utxoValidation: false,
+      });
+
+      const nextRoundTransactions = [];
+
+      for (let i = 0; i < dryRunResults.dryRun.length; i++) {
+        const currentResultIndex = transactionsToProcess[i];
+        const { receipts: rawReceipts, status } = dryRunResults.dryRun[i];
+        results[currentResultIndex].receipts = rawReceipts.map(processGqlReceipt);
+        results[currentResultIndex].dryrunStatus = status;
+
+        const { missingOutputVariables, missingOutputContractIds } = getReceiptsWithMissingData(
+          results[currentResultIndex].receipts
+        );
+        const hasMissingOutputs =
+          missingOutputVariables.length > 0 || missingOutputContractIds.length > 0;
+
+        const requestToProcess = allRequests[currentResultIndex];
+
+        if (hasMissingOutputs && requestToProcess?.type === TransactionType.Script) {
+          results[currentResultIndex].outputVariables += missingOutputVariables.length;
+          requestToProcess.addVariableOutputs(missingOutputVariables.length);
+          missingOutputContractIds.forEach(({ contractId }) => {
+            requestToProcess.addContractInputAndOutput(Address.fromString(contractId));
+            results[currentResultIndex].missingContractIds.push(contractId);
+          });
+
+          const { maxFee } = await this.estimateTxGasAndFee({
+            transactionRequest: requestToProcess,
+          });
+          requestToProcess.maxFee = maxFee;
+
+          // Prepare for the next round of dry run
+          serializedTransactionsMap.set(
+            currentResultIndex,
+            hexlify(requestToProcess.toTransactionBytes())
+          );
+          nextRoundTransactions.push(currentResultIndex);
+
+          allRequests[currentResultIndex] = requestToProcess;
+        }
+      }
```

Comment:
```suggestion
      for (let i = 0; i < dryRunResults.dryRun.length; i++) {
        const requestIdx = transactionsToProcess[i];
        const { receipts: rawReceipts, status } = dryRunResults.dryRun[i];
        const result = results[requestIdx];
        result.receipts = rawReceipts.map(processGqlReceipt);
        result.dryrunStatus = status;

        const { missingOutputVariables, missingOutputContractIds } = getReceiptsWithMissingData(
          result.receipts
        );
        const hasMissingOutputs =
          missingOutputVariables.length > 0 || missingOutputContractIds.length > 0;

        const request = allRequests[requestIdx];

        if (hasMissingOutputs && request?.type === TransactionType.Script) {
          result.outputVariables += missingOutputVariables.length;
          request.addVariableOutputs(missingOutputVariables.length);
          missingOutputContractIds.forEach(({ contractId }) => {
            request.addContractInputAndOutput(Address.fromString(contractId));
            result.missingContractIds.push(contractId);
          });

          const { maxFee } = await this.estimateTxGasAndFee({
            transactionRequest: request,
          });
          request.maxFee = maxFee;

          // Prepare for the next round of dry run
          serializedTransactionsMap.set(requestIdx, hexlify(request.toTransactionBytes()));
          nextRoundTransactions.push(requestIdx);
        }
      }
```

A bit of variable renaming and the like I did while trying to understand what's happening.

Reply:
https://github.com/FuelLabs/fuels-ts/pull/2021/commits/2a91c098b42b513fb4459f1ff9367ead1545201a